app/scripts/controllers/clientsmodal.js

```diff
@@ -8,23 +8,90 @@ angular.module('openhimConsoleApp')
 
-    // object for the taglist roles
-    $scope.taglistClientRoleOptions = [];
-
     // object to store temp values like password (not associated with schema object)
     $scope.temp = {};
-
-    // get the roles for the client taglist option
-    Api.Clients.query(function(clients){
-      angular.forEach(clients, function(client){
-        angular.forEach(client.roles, function(role){
-          if ( $scope.taglistClientRoleOptions.indexOf(role) === -1 ){
-            $scope.taglistClientRoleOptions.push(role);
-          }
-        });
+    
+    $scope.roles = {};
+    $scope.formData = {};
+    $scope.formData.assigned = {};
+    $scope.formData.newClientRole = null;
+    
+    var checkAssignedRoles = function () {
+      for(var i = 0; i<$scope.client.roles.length; i++) {
+        $scope.formData.assigned[$scope.client.roles[i]] = true;
+      }
+    };
+    
+    $scope.$watch('client', function() {
+      Api.Roles.query(function(roles) {
+        $scope.roles = roles;
+        if($scope.client.name) {
+          checkAssignedRoles();
+        } else {
+          $scope.client.roles = [];
+        }
       });
-    },
-    function(){ /* server error - could not connect to API to get clients */  });
+    });
+    
+    var removeRole = function(roleName) {
+      var index = -1;
+      for(var i = 0; i<$scope.client.roles.length; i++) {
+         if ($scope.client.roles[i] === roleName) {
+             index = i;
+             break;
+         }
+      }
+      $scope.client.roles.splice(index, 1);
+    };
+    
+    $scope.toggleAssignedRoles = function(role) {
+      if($scope.formData.assigned[role]) {
+        $scope.formData.assigned[role] = false;
+        removeRole(role);
+      } else {
+        $scope.formData.assigned[role] = true;
+        $scope.client.roles.push(role);
+      }
+    };
+    
+    var isDuplicateRole = function(role) {
+      var isDuplicate = false;
+      for(var i = 0; i<$scope.roles.length; i++) {
+        if($scope.roles[i].name === role) {
+          isDuplicate = true;
+        }
+      }
+      return isDuplicate;
+    };
+    
+    Api.Clients.query(function(clients) {
+      $scope.clients = clients;
+    });
+    
+    var isClient = function(role) {
+      var isClient = false;
+      for(var i = 0; i<$scope.clients.length; i++) {
+        if($scope.clients[i].clientID === role) {
+          isClient = true;
+        }
+      }
+      return isClient;
+    };
+    
+    $scope.createNewRole = function() {
+      var newRole = $scope.formData.newClientRole;
+      if(newRole) {
+        if(isDuplicateRole(newRole) || isClient(newRole)) {
+          $scope.formData.duplicateNewRole = true;
+        } else {
+          $scope.formData.duplicateNewRole = false;
+          $scope.client.roles.push(newRole);
+          $scope.roles.push({name: newRole});
+          $scope.formData.assigned[newRole] = true;
+          $scope.formData.newClientRole = null;
+        }
+      }
+    };
```

Comment:
A lot of this code looks very similar to the code in `channelsmodal.js` above. If it is repeated (it's hard to tell just by looking at it) then it would be much better to split this functions out into another file that is shared between these controllers. Otherwise, this is breaking the DRY (Don't repeat yourself) programming principle. Would it be possible to split it out?